File: src/main/java/io/github/ajoz/workshop/fp/java/tools/seq/MapFlow.java
package io.github.ajoz.workshop.fp.java.tools.seq;

import io.github.ajoz.workshop.fp.java.tools.Function1;
import io.github.ajoz.workshop.fp.java.tools.Try;

public final class MapFlow<A, B> implements Flow<B> {
    private final Flow<A> upstream;
    private final Function1<? super A, ? extends B> mapper;

    public MapFlow(final Flow<A> upstream,
                   final Function1<? super A, ? extends B> mapper) {
        this.upstream = upstream;
        this.mapper = mapper;
    }

    @Override
    public Try<B> next() {
        return upstream.next().map(mapper);
    }
}

File: kotlin/tools/src/main/kotlin/io/github/ajoz/workshop/fp/tools/flow/MapFlow.kt
package io.github.ajoz.workshop.fp.tools.flow

import io.github.ajoz.workshop.fp.tools.control.Try

/**
 *
 */
fun <A, B> Flow<A>.map(function: (A) -> B): Flow<B> {
    class MapFlow<A, B>(val upstream: Flow<A>,
                        val mapper: (A) -> B) : Flow<B> {

        override fun next(): Try<B> =
                upstream.next().map(mapper)
    }

    return MapFlow(this, function)
}
